gc/base/TLHAllocationInterface.cpp

```diff
@@ -200,21 +200,17 @@ MM_TLHAllocationInterface::allocateObject(MM_EnvironmentBase *env, MM_AllocateDe
 
 	}
 
-	if (NULL != result) {
-		uintptr_t sizeInBytesAllocated = allocDescription->getContiguousBytes();
-		/* Increment by bytes allocated */
-		env->_traceAllocationBytes += sizeInBytesAllocated;
-
-		if (!allocDescription->isCompletedFromTlh()) {
+	if (NULL != result && !allocDescription->isCompletedFromTlh()) {
 #if defined(OMR_GC_OBJECT_ALLOCATION_NOTIFY)
-			env->objectAllocationNotify((omrobjectptr_t)result);
+		env->objectAllocationNotify((omrobjectptr_t)result);
 #endif /* OMR_GC_OBJECT_ALLOCATION_NOTIFY */
-			_stats._allocationBytes += sizeInBytesAllocated;
-			_stats._allocationCount += 1;
-		}
+		_stats._allocationBytes += allocDescription->getContiguousBytes();
+		_stats._allocationCount += 1;
 	}
-	env->_oolTraceAllocationBytes += (_stats.bytesAllocated() - _bytesAllocatedBase); /* Increment by bytes allocated */
 
+	uintptr_t sizeInBytesAllocated = (_stats.bytesAllocated(true) - _bytesAllocatedBase);
+	env->_oolTraceAllocationBytesForTracepoint += sizeInBytesAllocated; /* Increment by bytes allocated for Trace */
+	env->_traceAllocationBytesForHook += sizeInBytesAllocated; /* Increment by bytes allocated for Hook */
```

Comment:
This is critical part to understand. You are saying that 
a) if we don't force any extra OOL allocations (due to sampling rate being < TLH size), that bytesAllocated(true) is equal to what bytesAllocated() used to be. Or in other words, that _tlhAllocatedFresh - _tlhDiscardedBytes == _tlhAllocatedUsed? Did you verify that - is it always true or just approximatelly true?
b), if we do force   extra OOL allocations (due to sampling rate being < TLH size), that SUM of all bytesAllocated(true) will still be equal to what bytesAllocated() used to be.

Reply:
I think that there might be a difference in behavior since _tlhAllocatedFresh was taken _after_ TLH was refreshed (hence it was for next TLH, what seems like not completely correct), and _tlhDiscardedBytes was taken _before_ TLH was refreshed (for previous TLH). On the other side _tlhAllocatedUsed is correctly for previous TLH. 
In a long run, when things are averaged _tlhAllocatedFresh - _tlhDiscardedBytes == _tlhAllocatedUsed might hold true, but strictly speaking, I suspect it does not hold true.

Reply:
Should we do special handling in case of dual TLH support? It was not relevant before (TLH allocation was disabled essentially so did not matter which one). But looks like it becomes relevant now. Is it correct? The simple solution can be disabling of non-zeroed TLH in case of instrumentation

Reply:
_tlhAllocatedUsed is the amount of memory has been flushed, "_tlhAllocatedFresh - _tlhDiscardedBytes == _tlhAllocatedUsed" is true when the TLH is just right before refreshing, after refreshing _tlhAllocatedFresh would be added with new TLH size.
bytesAllocated(true) try to return the size for all of allocated bytes excludes any bytes in current TLH. naming of _tlhAllocatedUsed might be too confuse.